File: ChineseCheckers/app/src/main/java/com/chinesecheckers/game/BoardDefinition.kt
package com.chinesecheckers.game

import kotlin.math.abs
import kotlin.math.sqrt

data class BoardPos(val row: Int, val col: Int)

object Board {
    val allPositions: Set<BoardPos> by lazy { buildPositions() }

    val player1Home: Set<BoardPos> by lazy {
        allPositions.filter { it.row in 0..3 }.toSet()
    }

    val player2Home: Set<BoardPos> by lazy {
        allPositions.filter { it.row in 13..16 }.toSet()
    }

    private fun buildPositions(): Set<BoardPos> {
        val positions = mutableSetOf<BoardPos>()

        // Top triangle rows 0–3 (Player 1 home)
        // Row r: (r+1) holes starting at col (12 - r), step 2
        for (row in 0..3) {
            val count = row + 1
            val startCol = 12 - row
            for (i in 0 until count) {
                positions += BoardPos(row, startCol + i * 2)
            }
        }

        // Middle section rows 4–12
        // Row r: (9 + |r-8|) holes starting at col (4 - |r-8|), step 2
        for (row in 4..12) {
            val dist = abs(row - 8)
            val count = 9 + dist
            val startCol = 4 - dist
            for (i in 0 until count) {
                positions += BoardPos(row, startCol + i * 2)
            }
        }

        // Bottom triangle rows 13–16 (Player 2 home)
        // Row r: (17-r) holes starting at col (r-4), step 2
        for (row in 13..16) {
            val count = 17 - row
            val startCol = row - 4
            for (i in 0 until count) {
                positions += BoardPos(row, startCol + i * 2)
            }
        }

        return positions
    }

    fun getNeighbors(pos: BoardPos): List<BoardPos> {
        val candidates = listOf(
            BoardPos(pos.row, pos.col - 2),
            BoardPos(pos.row, pos.col + 2),
            BoardPos(pos.row - 1, pos.col - 1),
            BoardPos(pos.row - 1, pos.col + 1),
            BoardPos(pos.row + 1, pos.col - 1),
            BoardPos(pos.row + 1, pos.col + 1)
        )
        return candidates.filter { it in allPositions }
    }

    fun jumpTarget(from: BoardPos, over: BoardPos): BoardPos =
        BoardPos(2 * over.row - from.row, 2 * over.col - from.col)

    // Maps doubled-coordinate position to screen pixel offset
    fun toScreen(pos: BoardPos, cellSize: Float, originX: Float, originY: Float): Pair<Float, Float> {
        val x = originX + pos.col * cellSize * 0.5f
        val y = originY + pos.row * cellSize * sqrt(3.0).toFloat() * 0.5f
        return x to y
    }
}

File: ChineseCheckersIOS/ChineseCheckersIOS/Game/BoardDefinition.swift
import Foundation
import CoreGraphics

// Doubled hex coordinates — row + col always same parity.
struct BoardPos: Hashable, Equatable, Codable {
    let row: Int
    let col: Int
}

enum Board {
    static let allPositions: Set<BoardPos> = buildPositions()
    static let player1Home:  Set<BoardPos> = Set(allPositions.filter { $0.row <= 3  })
    static let player2Home:  Set<BoardPos> = Set(allPositions.filter { $0.row >= 13 })

    private static func buildPositions() -> Set<BoardPos> {
        var set = Set<BoardPos>()
        // Top triangle rows 0–3
        for row in 0...3 {
            let start = 12 - row
            for i in 0...row { set.insert(.init(row: row, col: start + i * 2)) }
        }
        // Middle rows 4–12
        for row in 4...12 {
            let dist = abs(row - 8)
            let count = 9 + dist
            let start = 4 - dist
            for i in 0..<count { set.insert(.init(row: row, col: start + i * 2)) }
        }
        // Bottom triangle rows 13–16
        for row in 13...16 {
            let count = 17 - row
            let start = row - 4
            for i in 0..<count { set.insert(.init(row: row, col: start + i * 2)) }
        }
        return set
    }

    static func neighbors(of pos: BoardPos) -> [BoardPos] {
        [
            .init(row: pos.row,     col: pos.col - 2),
            .init(row: pos.row,     col: pos.col + 2),
            .init(row: pos.row - 1, col: pos.col - 1),
            .init(row: pos.row - 1, col: pos.col + 1),
            .init(row: pos.row + 1, col: pos.col - 1),
            .init(row: pos.row + 1, col: pos.col + 1),
        ].filter { allPositions.contains($0) }
    }

    static func jumpTarget(from: BoardPos, over: BoardPos) -> BoardPos {
        .init(row: 2 * over.row - from.row, col: 2 * over.col - from.col)
    }

    // Screen coordinates using doubled-coord spacing
    static func screenPoint(pos: BoardPos, cellSize: CGFloat, origin: CGPoint) -> CGPoint {
        CGPoint(
            x: origin.x + CGFloat(pos.col) * cellSize * 0.5,
            y: origin.y + CGFloat(pos.row) * cellSize * sqrt(3.0) / 2.0
        )
    }
}
